include/ghex/unstructured/communication_object_ipr.hpp

```diff
@@ -0,0 +1,561 @@
+                    template<typename Function>
+                    struct field_info {
+                            using index_container_type = typename pattern_type::map_type::mapped_type;
+                            Function call_back;
+                            const index_container_type* index_container;
+                            std::size_t offset;
+                            void* field_ptr;
+                    };
+
+                    /** @brief Holds serial buffer memory and meta information associated with it.
+                     * @tparam Vector contiguous buffer memory type
+                     * @tparam Function pack function pointer type (not used for unpacking)*/
+                    template<class Vector, class Function>
+                    struct buffer {
+                            using field_info_type = field_info<Function>;
+                            address_type address;
+                            int tag;
+                            Vector buffer;
+                            std::size_t size;
+                            std::vector<field_info_type> field_infos;
+                            cuda::stream m_cuda_stream;
+                    };
+
+                    /** @brief Message-like wrapper over in-place receive memory*/
+                    class ipr_message {
+
+                        public:
+
+                            using byte = unsigned char;
+                            using value_type = byte;
+
+                        private:
+
+                            value_type* m_ipr_ptr;
+                            std::size_t m_size;
+
+                        public:
+
+                            ipr_message() = default;
+                            ipr_message(value_type* ipr_ptr) : m_ipr_ptr{ipr_ptr}, m_size{0} {}
+                            ipr_message(value_type* ipr_ptr, const std::size_t size) : m_ipr_ptr{ipr_ptr}, m_size{size} {}
+                            const value_type* data() const { return m_ipr_ptr; }
+                            value_type* data() { return m_ipr_ptr; }
+                            std::size_t size() const { return m_size; }
+                            void resize(const std::size_t size) { m_size = size; }
+
+                    };
+
+                    /** @brief Equivalent of buffer for in-place receive
+                     * Vector = ipr_message, Function = std::function<void()>;
+                     * not a specialization, because it is no longer a buffer*/
+                    struct recv_ipr_info {
+                            using field_info_type = field_info<std::function<void()>>;
+                            address_type address;
+                            int tag;
+                            ipr_message message;
+                            std::size_t size;
+                            std::vector<field_info_type> field_infos;
+                            cuda::stream m_cuda_stream;
```

Comment:
Same comment here on the naming convention

Reply:
as above

Reply:
Yep